closed/src/java.base/share/classes/com/ibm/security/criu/DigestBase.java

```diff
@@ -0,0 +1,283 @@
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * ===========================================================================
+ */
+
+package com.ibm.security.criu;
+
+import java.security.MessageDigestSpi;
+import java.security.DigestException;
+import java.security.ProviderException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.ArrayList;
+
+import sun.security.action.GetPropertyAction;
+
+/**
+ * Common base message digest implementation for the Sun provider.
+ * It implements all the JCA methods as suitable for a Java message digest
+ * implementation of an algorithm based on a compression function (as all
+ * commonly used algorithms are). The individual digest subclasses only need to
+ * implement the following methods:
+ *
+ *  . abstract void implCompress(byte[] b, int ofs);
+ *  . abstract void implDigest(byte[] out, int ofs);
+ *  . abstract void implReset();
+ *
+ * See the inline documentation for details.
+ *
+ * @since   1.5
+ * @author  Andreas Sterbenz
+ */
+abstract class DigestBase extends MessageDigestSpi implements Cloneable {
+
+    // one element byte array, temporary storage for update(byte)
+    private byte[] oneByte;
+
+    // algorithm name to use in the exception message
+    private final String algorithm;
+    // length of the message digest in bytes
+    private final int digestLength;
+
+    // size of the input to the compression function in bytes
+    private final int blockSize;
+    // buffer to store partial blocks, blockSize bytes large
+    // Subclasses should not access this array directly except possibly in their
+    // implDigest() method. See MD5.java as an example.
+    byte[] buffer;
+    // offset into buffer
+    private int bufferOffset;
+
+    // number of bytes processed so far. subclasses should not modify
+    // this value.
+    // also used as a flag to indicate reset status
+    // -1: need to call engineReset() before next call to update()
+    //  0: is already reset
+    long bytesProcessed;
+
+    private static final ArrayList<DigestBase> digestList = new ArrayList<>();
+
+    /** Tracing for CRIUSEC */
+    private static final boolean debug = Boolean.parseBoolean(GetPropertyAction.privilegedGetProperty
+            ("enable.j9internal.checkpoint.security.api.debug", "false"));
+
+    /**
+     * This method is required to ensure that no state is maintained when checkpointing.
+     */
+    public static int resetDigests() {
+        if (debug) {
+            System.out.println("Clearing all digests ...");
+        }
+        synchronized (digestList) {
+            for (DigestBase digest : digestList) {
+                digest.engineReset();
+            }
+        }
+        return 0;
+    }
```

Comment:
changed to `void`

Reply:
I don't see any uses of this method: where will it be used?

Reply:
This method is used in the PR [eclipse-openj9/openj9#14824](https://github.com/eclipse-openj9/openj9/pull/14824) to reset the security digests after checkpoint so that no state is maintained.

Reply:
I don't see a use of this method in 14824; the nearest I found was this:
```
    com.ibm.security.criu.SHA.resetDigests();
```
But the method is not defined in `SHA`.

Reply:
The reference to a static method should be via the class that declares it (`DigestBase`, not `SHA`).

Reply:
Of course, `DigestBase` will need to be `public` for this to work.